subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c

```diff
@@ -0,0 +1,1554 @@
+			return SENML_JSON_FLOAT_VALUE_ATTRIBUTE;
+		}
+
+	} else if (fd->name_len == 2) {
+		if (attrbute_name[0] == 'b') {
+			if (attrbute_name[1] == 'n') {
+				return SENML_JSON_BASE_NAME_ATTRIBUTE;
+			} else if (attrbute_name[1] == 't') {
+				return SENML_JSON_BASE_TIME_ATTRIBUTE;
+			}
+		} else if (attrbute_name[0] == 'v') {
+			if (attrbute_name[1] == 'b') {
+				return SENML_JSON_BOOLEAN_VALUE_ATTRIBUTE;
+			} else if (attrbute_name[1] == 'd') {
+				return SENML_JSON_OPAQUE_VALUE_ATTRIBUTE;
+			} else if (attrbute_name[1] == 's') {
+				return SENML_JSON_STRING_VALUE_ATTRIBUTE;
+			}
+		}
+	} else if (fd->name_len == 3) {
+		if (attrbute_name[0] == 'v' && attrbute_name[1] == 'l' && attrbute_name[2] == 'o') {
+			return SENML_JSON_OBJ_LINK_VALUE_ATTRIBUTE;
+		}
+	}
+
+	return SENML_JSON_UNKNOWN_ATTRIBUTE;
+}
+
+/* Parse SenML attribute & value pairs  */
+static int json_next_token(struct lwm2m_input_context *in, struct json_in_formatter_data *fd)
+{
+	uint8_t cont, c = 0;
+	bool escape = false;
+	struct lwm2m_senml_json_context *block_ctx;
+
+	(void)memset(fd, 0, sizeof(*fd));
+	cont = 1U;
+	block_ctx = seml_json_context_get(in->block_ctx);
+
+	if (block_ctx && block_ctx->json_flags) {
+		/* Store from last sequence */
+		fd->json_flags = block_ctx->json_flags;
+		block_ctx->json_flags = 0;
+	}
+
+	/* We will be either at start, or at a specific position */
+	while (in->offset < in->in_cpkt->offset && cont) {
+		fd->offset = in->offset;
+		if (buf_read_u8(&c, CPKT_BUF_READ(in->in_cpkt), &in->offset) < 0) {
+			break;
+		}
+
+		if (c == '\\') {
+			escape = true;
+			/* Keep track of the escape codes */
+			json_add_char(in, fd);
+			continue;
+		}
+
+		switch (c) {
+		case '[':
+			if (!escape) {
+				fd->json_flags |= T_OBJECT_BEGIN;
+				cont = 0U;
+			} else {
+				json_add_char(in, fd);
+			}
+			break;
+
+		case '}':
+		case ']':
+			if (!escape) {
+				fd->json_flags |= T_OBJECT_END;
+				cont = 0U;
+			} else {
+				json_add_char(in, fd);
+			}
+			break;
+
+		case '{':
+			if (!escape) {
+				fd->json_flags |= T_OBJECT_BEGIN;
+			} else {
+				json_add_char(in, fd);
+			}
+
+			break;
+
+		case ',':
+			if (!escape) {
+				cont = 0U;
+			} else {
+				json_add_char(in, fd);
+			}
+
+			break;
+
+		case '"':
+			if (!escape) {
+				if (fd->json_flags & T_STRING_BEGIN) {
+					fd->json_flags &= ~T_STRING_BEGIN;
+					fd->json_flags |= T_STRING_END;
+				} else {
+					fd->json_flags &= ~T_STRING_END;
+					fd->json_flags |= T_STRING_BEGIN;
+				}
+			} else {
+				json_add_char(in, fd);
+			}
+
+			break;
+
+		case ':':
+			if (!escape) {
+				fd->json_flags &= ~T_STRING_END;
+				fd->json_flags |= T_VALUE;
+			} else {
+				json_add_char(in, fd);
+			}
+
+			break;
+
+		/* ignore whitespace */
+		case ' ':
+		case '\n':
+		case '\t':
+			if (!(fd->json_flags & T_STRING_BEGIN)) {
+				break;
+			}
+
+			__fallthrough;
+
+		default:
+			json_add_char(in, fd);
+		}
+
+		if (escape) {
+			escape = false;
+		}
+	}
```

Comment:
This code is almost identical to this:
https://github.com/zephyrproject-rtos/zephyr/blob/ccf51e3cd3ae455d94fbf3706e57c62cc3fc318b/subsys/net/lib/lwm2m/lwm2m_rw_json.c#L154-L248

and the same is probably true for other parts of this file.

Furthermore we have a general json library here, for parsing of json:
https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/os/json.c

Controlled with [CONFIG_JSON_LIBRARY](https://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_JSON_LIBRARY.html)

Why do we need so many versions / copies of json parsers ?
It increases maintenance burden, not to mention, if one library fixes a bug, that very same bug has a risk of being present also in another lib.

Could you please align to a single implementation for json.

Reply:
I looking for option to use CONFIG_JSON_LIBRARY. I need to study what need I need to change if I start looking for. I propose that I integrated SenML Json part first to use CONFIG_JSON_LIBRARY.

Reply:
Followup task created: https://github.com/zephyrproject-rtos/zephyr/issues/43103